### ISSUE TITLE
Settings --delete-trailing-commas --want-trailing-commas=b accidentally deletes comma before wrapping closing parenthesis

### ISSUE THREAD

**Issue by dboehmer** [perltidy/perltidy#156]
I just found that the combination of `--delete-trailing-commas` and `--want-trailing-commas=b` might delete commas that I want not be deleted. It seems to me like the comma is deleted first before the closing parenthesis is wrapped to the next line. Can this be fixed?

```bash
perltidy --noprofile --delete-trailing-commas --want-trailing-commas=b <<EOT
f(
    a => 1,
    b => 2, );
EOT
```

yields to the output:

```perl
f(
    a => 1,
    b => 2
);
```

This is stable though as expected:

```bash
perltidy --noprofile --delete-trailing-commas --want-trailing-commas=b <<EOT
f(
    a => 1,
    b => 2,
);
EOT
```

This issue isn’t about tidying code in tests/CI but about typing code lazily and making Perl-Tidy clear it up. I have a hotkey in my Vim config to tidy the file and expected Perl-Tidy to transform the input at the top to this stable form.

I am using the latest version:

```console
$ cpanm Perl::Tidy
Perl::Tidy is up to date. (20240511)
```

**Comment by shancock9** [comment]
The operations which add and delete commas are done very early and make decisions based on the starting line breaks. 

Although the order of operations can't be changed, a way to fix this problem is to add parameters
``--add-trailing-commas`` ( or ``-atc``) and ``--converge`` (or ``-conv``).  

The ``-atc`` parameter gives permission to add commas which do not match the style.

When the ``--converge`` parameter is set, and a change in line break locations is detected,
another formatting pass is made.  On this second pass, if the ``--add-trailing-commas`` parameter is set, 
the comma will be added back.  So on the original snippet the result will be as expected:

```
# perltidy --npro -dtc -atc -wtc=b -conv
f(
    a => 1,
    b => 2,
);

The only disadvantage of using ``--converge`` is that it can increase run time.
But for small files that will not be noticable.

**Comment by dboehmer** [comment]
Unfortunately I don’t enforce a specific trailing comma rule for lists. For some lists I don’t add one, e.g. for method calls with a fixed number of arguments, others lists get a trailing comma. So I don’t want to enable `--add-trailing-commas`.

I can understand why Perl-Tidy works differently and it’s probably currently impossible to fix this.

**Comment by shancock9** [comment]
Another option would be to just use ``--delete-trailing-commas`` occasionally, if needed.
Commas would remain unchanged until then.

**Comment by shancock9** [comment]
Another option would be to write a macro to run perltidy twice, waiting until the second pass to turn on -dtc. Something like

```
perltidy -st -ndtc | perltidy -dtc -wtc=b
```

The ``-ndtc`` on the command line would override any -dtc in a ``perltidyrc`` file.

**Comment by shancock9** [comment]
Tagged version 20240511.07 at

https://github.com/perltidy/perltidy/tags

has an update which fixes this problem if iterations are requested.
For the example: 

```
# perltidy --delete-trailing-commas --want-trailing-commas=b --converge
f(
    a => 1,
    b => 2,
);
```

It is not necessary to use ``--add-trailing-commas``.  It works by waiting
until the second iteration to make changes to commas.

**Comment by Julien-Elie** [comment]
I see you added `-dtco` (`--delay-trailing-comma-operations`) besides the suggestion of using `--converge`.  Is this new option intended for that use?  (It is not documented in the latest 20240511.08 tag.)

By the way, I have checked for possible other undocumented options and found out the following ones that perltidy recognizes but are not documented: `fmt dtco syn pscf lal lsl pvl exp fll scl v`.  Maybe they should be officially documented?  (I have not checked whether the list of "switches which may be negated" is still accurate in the documentation.)

**Comment by shancock9** [comment]
Many thanks for the feedback on version 20240511.08, it is really helpful.

I checked the undocumented abbreviations and found the following:

``-dtco, --delay-trailing-comma-operations`` is a new parameter which has now been added.

``-syn, --check-syntax`` and ``-pscf, --perl-syntax-check-flags`` are relics
from when perltidy did syntax checking. They do nothing but are still accepted 
to avoid breaking ancient perltidyrc files.

``-lal, --look-for-autoloader``, ``-lsl --look-for-selfloader``, and ``-pvl--pass-version-line``, 
were only defined by their negatives, ``-nlal`` etc. I've added the positive definitions.

``-exp, --experimental`` was a relic from past testing which does nothing, so I've removed it.

``-fll, --fuzzy-line-length`` and ``scl --short-concatenation-item-length`` are
line length tolerances used for stress-testing. They are not otherwise useful and can remain undocumented.

``-fmt,  --format`` is the main formatting switch normally set with one of two synonyms: 
``--format=tidy`` or simply ``--tidy``, the default, and
``--format=html``, or simply ``--html``.  I've added an entry for it.

``-v, --version`` is surprisingly missing and has been added. ``perltidy -v``
dumps the version number and exits. 

The end table of parameters which may be negated is automatically generated
when cpan releases are made, so it should be accurate except perhaps for very
recent changes.

I've just added tagged version 20240511.09 which updates ``--warn-variable-types``
to include a new check type 'c' for unused constants (defined with ``use constant``).
This should be close to the next CPAN release, which should be in a few weeks. 
Thanks again.
